### PR TITLE
builtins: make pg_get_indexdef handle expression indexes

### LINKED ISSUE #94690: SHOW INDEX/ES stmts information about computed indexes is misleading
**Is your feature request related to a problem? Please describe.**

Imagine the following schema:

```sql
CREATE TABLE foo (id string PRIMARY key, json JSON);
```

With the following index:

```sql
CREATE INDEX foo__json__bar ON foo (id, (json->>'bar'));
```

If I use my IDE to get information about which JSON property I am indexing, I get this:

imagen

Just `crdb_internal_idx_expr` which is not very useful.

Which is indeed the same information I get if I execute the following stmt:

```sql
SHOW INDEXES FROM foo
```

```
  table_name |   index_name   | non_unique | seq_in_index |      column_name       | direction | storing | implicit | visible
-------------+----------------+------------+--------------+------------------------+-----------+---------+----------+----------
  foo        | foo__json__bar |     t      |            1 | id                     | ASC       |    f    |    f     |    t
  foo        | foo__json__bar |     t      |            2 | crdb_internal_idx_expr | ASC       |    f    |    f     |    t
  foo        | foo_pkey       |     f      |            1 | id                     | ASC       |    f    |    f     |    t
  foo        | foo_pkey       |     f      |            2 | json                   | N/A       |    t    |    f     |    t
(4 rows)
```

To get useful information about the index, I have to use:

```sql
SHOW CREATE TABLE foo
```

```
  table_name |                        create_statement
-------------+-----------------------------------------------------------------
  foo        | CREATE TABLE public.foo (
             |     id STRING NOT NULL,
             |     json JSONB NULL,
             |     CONSTRAINT foo_pkey PRIMARY KEY (id ASC),
             |     INDEX foo__json__bar (id ASC, (json->>'bar':::STRING) ASC)
             | )
(1 row)
```

Where I can see the useful piece of information I was looking for, but it seems a bit misleading having to use `SHOW CREATE TABLE` to gather information about an index. IMHO `SHOW INDEXES` returning it would be awesome, and possibly the IDEs could benefit from this feature as well.

**Describe the solution you'd like**

Return the json/computed column information in the `SHOW INDEXES` stmt:

>  INDEX foo__json__bar (id ASC, (json->>'bar':::STRING) ASC)

**Describe alternatives you've considered**

I can use `SHOW CREATE TABLE foo` for now, but it's not very intuitive. 

**Additional context**

Thank you Shaun (Cockroach Labs) from CockroachDB Community Slack for the useful tip!

Jira issue: CRDB-23086

Epic CRDB-17785

### PR DESCRIPTION
**Pull request by rafiss**
The 1st commit can be backported.

---

fixes https://github.com/cockroachdb/cockroach/issues/94690

Release note (bug fix): The pg_get_indexdef was fixed so that it shows
the expression used to define an expression-based index.  In addition,
the function was previously including columns stored by the index,
which was incorrect and has now been fixed.

---

builtins: implement pg_get_indexdef as UDF 

---

sql: add index expression to SHOW INDEXES

Release note (sql change): SHOW INDEXES will now show the expression
used to define an index, if one was used.

### CODE REVIEW

**Comment by Xiang-Gu** [comment]
Since this PR is going to close the issue, can we add a test for `SHOW INDEXES FROM t`?

**Comment by rafiss** [comment]
hm, i didn't actually change SHOW INDEXES in this PR, but let me try doing that

File: pkg/sql/sem/builtins/pg_builtins.go [hunk 1]
```diff
@@ -720,8 +666,31 @@ var pgBuiltins = map[string]builtinDefinition{
...
+			Volatility: volatility.Stable,
+		},
+		tree.Overload{
+			IsUDF:      true,
+			Types:      tree.ParamTypes{{Name: "index_oid", Typ: types.Oid}, {Name: "column_no", Typ: types.Int}, {Name: "pretty_bool", Typ: types.Bool}},
+			ReturnType: tree.FixedReturnType(types.String),
+			Body: `SELECT CASE
+    				WHEN $2 = 0 THEN defs.indexdef
+						WHEN $2 < 0 OR $2 > array_length(i.indkey, 1) THEN ''
+						WHEN i.indkey[$2-1] = 0 THEN (indexprs::STRING[])[array_position(array_positions(i.indkey, 0), $2)]
```
Comment (knz) [inline pkg/sql/sem/builtins/pg_builtins.go:684]: Like before this would benefit from an explanatory comment nearby.

**Review by knz — APPROVED** [review]
This is very nice. Thanks.
I assume the string "N/A" is also what postgres uses. I was surprised about that. It's perhaps useful to highlight in a comment that this is intentional.

I have a few curiosity driven questions below but they need not block this PR.

**Review by msirek — COMMENTED** [review]
Nice addition. I just have a couple non-blocking comments.

> Reviewed 5 of 5 files at r1, 2 of 2 files at r2, 23 of 23 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95413#-:-NMCihStBDN_zxpyiIK5:bogm2bt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @Xiang-Gu, and @ZhouXing19)

___
*[`pkg/sql/delegate/show_database_indexes.go` line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95413#-NMCgqI87Ifg5W89uB4x:-NMCgqI9DzMq9KZii-Vi:b-uhfgp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7f9d6e0279d872c0344502f266d354aba1a875/pkg/sql/delegate/show_database_indexes.go#L59)):*
> ```Go
>     JOIN %[1]s.pg_catalog.pg_class c_table ON c_table.relname = s.table_name
>     JOIN %[1]s.pg_catalog.pg_namespace n ON c.relnamespace = n.oid AND c_table.relnamespace = n.oid AND n.nspname = s.index_schema
>     JOIN %[1]s.pg_catalog.pg_index i ON i.indexrelid = c.oid AND i.indrelid = c_table.oid
> ```

It seems like there's an unnecessary join to pg_class here. It could be rewritten as:
```
FROM
    %[1]s.information_schema.statistics AS s
    JOIN %[1]s.pg_catalog.pg_class c ON c.relname = s.index_name OR c.relname = s.table_name
    JOIN %[1]s.pg_catalog.pg_namespace n ON c.relnamespace = n.oid AND n.nspname = s.index_schema
    JOIN %[1]s.pg_catalog.pg_index i ON i.indexrelid = c.oid
```
With this rewrite and a test case I tried with 9 indexes, average runtime dropped from about 50 ms to 35 ms.
___
*[`pkg/sql/delegate/show_table.go` line 161 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95413#-NMCi3Ho1yg33wuMZmKZ:-NMCi3Ho1yg33wuMZmK_:bo2qgnf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7f9d6e0279d872c0344502f266d354aba1a875/pkg/sql/delegate/show_table.go#L161)):*
> ```Go
>     JOIN %[4]s.pg_catalog.pg_class c_table ON c_table.relname = s.table_name
>     JOIN %[4]s.pg_catalog.pg_namespace n ON c.relnamespace = n.oid AND c_table.relnamespace = n.oid AND n.nspname = s.index_schema
>     JOIN %[4]s.pg_catalog.pg_index i ON i.indexrelid = c.oid AND i.indrelid = c_table.oid
> ```

Same as previous comment. The extra join to pg_class could be removed.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95413)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msirek, @Xiang-Gu, and @ZhouXing19)

___
*[`pkg/sql/delegate/show_database_indexes.go` line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95413#-NMCgqI87Ifg5W89uB4x:-NMCo_FI-70K_0dsLgWd:bs3u4li) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7f9d6e0279d872c0344502f266d354aba1a875/pkg/sql/delegate/show_database_indexes.go#L59)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

It seems like there's an unnecessary join to pg_class here. It could be rewritten as:
```
FROM
    %[1]s.information_schema.statistics AS s
    JOIN %[1]s.pg_catalog.pg_class c ON c.relname = s.index_name OR c.relname = s.table_name
    JOIN %[1]s.pg_catalog.pg_namespace n ON c.relnamespace = n.oid AND n.nspname = s.index_schema
    JOIN %[1]s.pg_catalog.pg_index i ON i.indexrelid = c.oid
```
With this rewrite and a test case I tried with 9 indexes, average runtime dropped from about 50 ms to 35 ms.

</blockquote></details>

those don't seem equivalent. it can lead to double-counting if there are two tables that have indexes with the same names. example:

```
root@localhost:26257/defaultdb> create table t (a int, b int, index idx (b));
CREATE TABLE

root@localhost:26257/defaultdb> create table t2 (a int, b int, index idx (a));
CREATE TABLE

root@localhost:26257/defaultdb> select relname, oid from pg_class where relname = 'idx';
  relname |    oid
----------+-------------
  idx     |  737994182
  idx     | 3428148196

root@localhost:26257/defaultdb> SELECT
                             -> table_name,
                             -> index_name,
                             -> index_schema,
                             -> non_unique::BOOL,
                             -> seq_in_index,
                             -> column_name,
                             -> direction,
                             -> storing::BOOL,
                             -> implicit::BOOL,
                             -> is_visible::BOOL AS visible
                             -> FROM
                             ->     information_schema.statistics AS s
                             ->     JOIN pg_catalog.pg_class c ON c.relname = s.index_name OR c.relname = s.table_name
                             ->     JOIN pg_catalog.pg_namespace n ON c.relnamespace = n.oid AND n.nspname = s.index_schema
                             ->     JOIN pg_catalog.pg_index i ON i.indexrelid = c.oid
                             -> WHERE
                             ->     table_catalog='defaultdb'
                             ->     AND table_schema='public'
                             ->     AND table_name='t'
                             -> ORDER BY
                             ->     1, 2, 4;
  table_name | index_name | index_schema | non_unique | seq_in_index | column_name | direction | storing | implicit | visible
-------------+------------+--------------+------------+--------------+-------------+-----------+---------+----------+----------
  t          | idx        | public       |     t      |            1 | b           | ASC       |    f    |    f     |    t
  t          | idx        | public       |     t      |            2 | rowid       | ASC       |    f    |    t     |    t
  t          | idx        | public       |     t      |            1 | b           | ASC       |    f    |    f     |    t
  t          | idx        | public       |     t      |            2 | rowid       | ASC       |    f    |    t     |    t
  t          | t_pkey     | public       |     f      |            1 | rowid       | ASC       |    f    |    f     |    t
  t          | t_pkey     | public       |     f      |            3 | b           | N/A       |    t    |    f     |    t
  t          | t_pkey     | public       |     f      |            2 | a           | N/A       |    t    |    f     |    t

root@localhost:26257/defaultdb> drop table t2;
DROP TABLE

-- after dropping t2, there's no more double counting
root@localhost:26257/defaultdb> SELECT
                             -> table_name,
                             -> index_name,
                             -> index_schema,
                             -> non_unique::BOOL,
                             -> seq_in_index,
                             -> column_name,
                             -> direction,
                             -> storing::BOOL,
                             -> implicit::BOOL,
                             -> is_visible::BOOL AS visible
                             -> FROM
                             ->     information_schema.statistics AS s
                             ->     JOIN pg_catalog.pg_class c ON c.relname = s.index_name OR c.relname = s.table_name
                             ->     JOIN pg_catalog.pg_namespace n ON c.relnamespace = n.oid AND n.nspname = s.index_schema
                             ->     JOIN pg_catalog.pg_index i ON i.indexrelid = c.oid
                             -> WHERE
                             ->     table_catalog='defaultdb'
                             ->     AND table_schema='public'
                             ->     AND table_name='t'
                             -> ORDER BY
                             ->     1, 2, 4;
  table_name | index_name | index_schema | non_unique | seq_in_index | column_name | direction | storing | implicit | visible
-------------+------------+--------------+------------+--------------+-------------+-----------+---------+----------+----------
  t          | idx        | public       |     t      |            1 | b           | ASC       |    f    |    f     |    t
  t          | idx        | public       |     t      |            2 | rowid       | ASC       |    f    |    t     |    t
  t          | t_pkey     | public       |     f      |            1 | rowid       | ASC       |    f    |    f     |    t
  t          | t_pkey     | public       |     f      |            3 | b           | N/A       |    t    |    f     |    t
  t          | t_pkey     | public       |     f      |            2 | a           | N/A       |    t    |    f     |    t
```

i'm open to hearing other ways of avoiding a join though. i added it because i needed it to properly filter the join to pg_index (`ON i.indexrelid = c.oid AND i.indrelid = c_table.oid`).

**Comment by rafiss** [comment]
I used "N/A" since that's what `SHOW INDEXES ...` uses for the `direction` when showing columns that are stored in the index. That comes directly from `information_schema.statistics`: https://github.com/cockroachdb/cockroach/blob/0775fcc3b3a71b47dc67a73f2caadbf4b586ec37/pkg/sql/information_schema.go#L1234

Neither `SHOW INDEXES` nor `information_schema.statistics` exists in PostgreSQL. (MySQL does have `information_schema.statistics`.) However, your comment reminded me that PostgreSQL does have:
```
postgres=# \dS+ i3
                       Index "public.i3"
 Column |  Type   | Key? | Definition | Storage | Stats target
--------+---------+------+------------+---------+--------------
 b      | integer | yes  | b          | plain   |
 expr   | integer | yes  | (a + c)    | plain   |
```

So now I think it would be nicer if we also use the term "definition" and just show the column name if there is no expression, like what `\dS+` does.

File: pkg/sql/logictest/testdata/logic_test/pg_builtins [hunk 2]
```diff
@@ -811,3 +811,24 @@ query T
...
+CREATE INDEX expr_idx ON test.public.expr_idx_tbl USING btree (id ASC, (json->>'bar'::STRING) ASC, length(id) ASC)
+
+query IT
+SELECT k, pg_get_indexdef('expr_idx'::regclass::oid, k, true) FROM generate_series(0,4) k ORDER BY k
+----
+0  CREATE INDEX expr_idx ON test.public.expr_idx_tbl USING btree (id ASC, (json->>'bar'::STRING) ASC, length(id) ASC)
+1  id
+2  (json->>'bar'::STRING)
+3  (length(id))
+4  ·
```
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/sql/logictest/testdata/logic_test/pg_builtins:834]: `.` is the empty string. pg_get_indexdef is supposed to return the empty string for any column index that is not an index key. (confusing that the word index is overloaded in that sentence)

File: pkg/sql/logictest/testdata/logic_test/builtin_function [hunk 3]
```diff
@@ -2259,7 +2259,7 @@ b
 query T
 SELECT pg_catalog.pg_get_indexdef((SELECT oid from pg_class WHERE relname='pg_indexdef_cols_idx'), 3, false)
 ----
-rowid
+·
```
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/sql/logictest/testdata/logic_test/builtin_function:2262]: `.` is what the test uses to represent the empty string. showing `rowid` in the result here was actually a bug. I've added that to the release note, and left a comment on the test to explain what is being tested.

File: pkg/sql/delegate/show_table.go [hunk 4]
```diff
@@ -137,26 +137,29 @@ SELECT
     non_unique::BOOL,
     seq_in_index,
     column_name,
+    CASE
+      WHEN seq_in_index <= 0 OR seq_in_index > array_length(i.indkey, 1) THEN 'N/A'
+      WHEN i.indkey[seq_in_index-1] = 0 THEN (indexprs::STRING[])[array_position(array_positions(i.indkey, 0), seq_in_index)]
```
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/sql/delegate/show_table.go:142]: adding

File: pkg/sql/delegate/show_database_indexes.go [hunk 5]
```diff
@@ -31,28 +31,33 @@ func (d *delegator) delegateShowDatabaseIndexes(
 SELECT
 	table_name,
 	index_name,
+	index_schema,
 	non_unique::BOOL,
 	seq_in_index,
 	column_name,
+    CASE
+      WHEN seq_in_index <= 0 OR seq_in_index > array_length(i.indkey, 1) THEN 'N/A'
+      WHEN i.indkey[seq_in_index-1] = 0 THEN (indexprs::STRING[])[array_position(array_positions(i.indkey, 0), seq_in_index)]
```
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/sql/delegate/show_database_indexes.go:40]: adding

File: pkg/sql/sem/builtins/pg_builtins.go (hunk 1 above)
Comment (rafiss) [reply pkg/sql/sem/builtins/pg_builtins.go:689]: added

**Comment by knz** [comment]
Indeed my question about N/A was driven by my work on #88061 so we get an unsurprising output on \d

**Comment by rafiss** [comment]
tftr!

bors r=knz

**Comment by lopezator** [comment]
Awesome!!! Thank you for dealing with this @rafiss :)

**Review by msirek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95413)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/delegate/show_database_indexes.go` line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95413#-NMCgqI87Ifg5W89uB4x:-NMFQEahEeM8_enL5rFE:bstfhy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7f9d6e0279d872c0344502f266d354aba1a875/pkg/sql/delegate/show_database_indexes.go#L59)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

those don't seem equivalent. it can lead to double-counting if there are two tables that have indexes with the same names. example:

```
root@localhost:26257/defaultdb> create table t (a int, b int, index idx (b));
CREATE TABLE

root@localhost:26257/defaultdb> create table t2 (a int, b int, index idx (a));
CREATE TABLE

root@localhost:26257/defaultdb> select relname, oid from pg_class where relname = 'idx';
  relname |    oid
----------+-------------
  idx     |  737994182
  idx     | 3428148196

root@localhost:26257/defaultdb> SELECT
                             -> table_name,
                             -> index_name,
                             -> index_schema,
                             -> non_unique::BOOL,
                             -> seq_in_index,
                             -> column_name,
                             -> direction,
                             -> storing::BOOL,
                             -> implicit::BOOL,
                             -> is_visible::BOOL AS visible
                             -> FROM
                             ->     information_schema.statistics AS s
                             ->     JOIN pg_catalog.pg_class c ON c.relname = s.index_name OR c.relname = s.table_name
                             ->     JOIN pg_catalog.pg_namespace n ON c.relnamespace = n.oid AND n.nspname = s.index_schema
                             ->     JOIN pg_catalog.pg_index i ON i.indexrelid = c.oid
                             -> WHERE
                             ->     table_catalog='defaultdb'
                             ->     AND table_schema='public'
                             ->     AND table_name='t'
                             -> ORDER BY
                             ->     1, 2, 4;
  table_name | index_name | index_schema | non_unique | seq_in_index | column_name | direction | storing | implicit | visible
-------------+------------+--------------+------------+--------------+-------------+-----------+---------+----------+----------
  t          | idx        | public       |     t      |            1 | b           | ASC       |    f    |    f     |    t
  t          | idx        | public       |     t      |            2 | rowid       | ASC       |    f    |    t     |    t
  t          | idx        | public       |     t      |            1 | b           | ASC       |    f    |    f     |    t
  t          | idx        | public       |     t      |            2 | rowid       | ASC       |    f    |    t     |    t
  t          | t_pkey     | public       |     f      |            1 | rowid       | ASC       |    f    |    f     |    t
  t          | t_pkey     | public       |     f      |            3 | b           | N/A       |    t    |    f     |    t
  t          | t_pkey     | public       |     f      |            2 | a           | N/A       |    t    |    f     |    t

root@localhost:26257/defaultdb> drop table t2;
DROP TABLE

-- after dropping t2, there's no more double counting
root@localhost:26257/defaultdb> SELECT
                             -> table_name,
                             -> index_name,
                             -> index_schema,
                             -> non_unique::BOOL,
                             -> seq_in_index,
                             -> column_name,
                             -> direction,
                             -> storing::BOOL,
                             -> implicit::BOOL,
                             -> is_visible::BOOL AS visible
                             -> FROM
                             ->     information_schema.statistics AS s
                             ->     JOIN pg_catalog.pg_class c ON c.relname = s.index_name OR c.relname = s.table_name
                             ->     JOIN pg_catalog.pg_namespace n ON c.relnamespace = n.oid AND n.nspname = s.index_schema
                             ->     JOIN pg_catalog.pg_index i ON i.indexrelid = c.oid
                             -> WHERE
                             ->     table_catalog='defaultdb'
                             ->     AND table_schema='public'
                             ->     AND table_name='t'
                             -> ORDER BY
                             ->     1, 2, 4;
  table_name | index_name | index_schema | non_unique | seq_in_index | column_name | direction | storing | implicit | visible
-------------+------------+--------------+------------+--------------+-------------+-----------+---------+----------+----------
  t          | idx        | public       |     t      |            1 | b           | ASC       |    f    |    f     |    t
  t          | idx        | public       |     t      |            2 | rowid       | ASC       |    f    |    t     |    t
  t          | t_pkey     | public       |     f      |            1 | rowid       | ASC       |    f    |    f     |    t
  t          | t_pkey     | public       |     f      |            3 | b           | N/A       |    t    |    f     |    t
  t          | t_pkey     | public       |     f      |            2 | a           | N/A       |    t    |    f     |    t
```

i'm open to hearing other ways of avoiding a join though. i added it because i needed it to properly filter the join to pg_index (`ON i.indexrelid = c.oid AND i.indrelid = c_table.oid`).
</blockquote></details>

I see. Maybe an EXISTS clause could be used to prevent duplicates:
```sql
SELECT
table_name,
index_name,
index_schema,
non_unique::BOOL,
seq_in_index,
column_name,
direction,
storing::BOOL,
implicit::BOOL,
is_visible::BOOL AS visible
FROM
    information_schema.statistics AS s
    JOIN pg_catalog.pg_namespace n ON n.nspname = s.index_schema
WHERE
    table_catalog='defaultdb'
    AND table_schema='public'
    AND table_name='t'
    AND EXISTS (SELECT 1 FROM pg_catalog.pg_class c, pg_catalog.pg_index i WHERE (c.relname = s.index_name OR c.relname = s.table_name) AND c.relnamespace = n.oid AND i.indexrelid = c.oid)
ORDER BY
    1, 2, 4;  
```

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95413)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/delegate/show_database_indexes.go` line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95413#-NMCgqI87Ifg5W89uB4x:-NMFU0lb2Afyij8T7iVI:bowr308) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7f9d6e0279d872c0344502f266d354aba1a875/pkg/sql/delegate/show_database_indexes.go#L59)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

I see. Maybe an EXISTS clause could be used to prevent duplicates:
```sql
SELECT
table_name,
index_name,
index_schema,
non_unique::BOOL,
seq_in_index,
column_name,
direction,
storing::BOOL,
implicit::BOOL,
is_visible::BOOL AS visible
FROM
    information_schema.statistics AS s
    JOIN pg_catalog.pg_namespace n ON n.nspname = s.index_schema
WHERE
    table_catalog='defaultdb'
    AND table_schema='public'
    AND table_name='t'
    AND EXISTS (SELECT 1 FROM pg_catalog.pg_class c, pg_catalog.pg_index i WHERE (c.relname = s.index_name OR c.relname = s.table_name) AND c.relnamespace = n.oid AND i.indexrelid = c.oid)
ORDER BY
    1, 2, 4;  
```

</blockquote></details>

ah, that works for the old query, but for the new one, we need to select from `pg_index`, so `pg_index` can't only be in an inner subquery 

```
root@localhost:26257/defaultdb> SELECT
                             -> table_name,
                             -> index_name,
                             -> index_schema,
                             -> non_unique::BOOL,
                             -> seq_in_index,
                             -> column_name,
                             -> direction,
                             -> storing::BOOL,
                             -> implicit::BOOL,
                             -> is_visible::BOOL AS visible,
                             -> CASE
                             ->   WHEN i.indkey[seq_in_index-1] = 0 THEN (indexprs::STRING[])[array_position(array_positions(i.indkey, 0), seq_in_index)]
                             ->   ELSE column_name
                             -> END AS definition
                             -> FROM
                             ->     information_schema.statistics AS s
                             ->     JOIN pg_catalog.pg_namespace n ON n.nspname = s.index_schema
                             -> WHERE
                             ->     table_catalog='defaultdb'
                             ->     AND table_schema='public'
                             ->     AND table_name='t'
                             ->     AND EXISTS (SELECT 1 FROM pg_catalog.pg_class c, pg_catalog.pg_index i WHERE (c.relname = s.index_name OR c.relname =
                             -> s.table_name) AND c.relnamespace = n.oid AND i.indexrelid = c.oid)
                             -> ORDER BY
                             ->     1, 2, 4;
ERROR: no data source matches prefix: i in this context
SQLSTATE: 42P01
```

**Pull request merged**